src/app/Dashboard/Charts/ChartCard.tsx

```diff
@@ -0,0 +1,360 @@
+
+  'Compiler Peak Time' = 24,
+
+  'Object Allocation Sample' = 38,
+}
+
+export function kindToId(kind: string): number {
+  return ChartKind[kind];
+}
+
+function getHeight(kind: string): number {
+  switch (kind) {
+    case 'Core Count':
+      return 180;
+    default:
+      return 380;
+  }
+}
+
+export const ChartCard: React.FC<ChartCardProps> = (props) => {
+  const serviceContext = React.useContext(ServiceContext);
+  const controllerContext = React.useContext(ChartContext);
+  const history = useHistory();
+  const addSubscription = useSubscriptions();
+  const [idx, setIdx] = React.useState(0);
+  const [key, setKey] = React.useState(Math.floor(Math.random()));
+  const [hasRecording, setHasRecording] = React.useState(false);
+  const [chartSrc, setChartSrc] = React.useState('');
+  const [dashboardUrl, setDashboardUrl] = React.useState('');
+
+  const updateKey = React.useCallback(() => {
+    setKey((prev) => {
+      let next = prev + 1;
+      if (next >= 10) {
+        next = 0;
+      }
+      return next;
+    });
+  }, [setKey]);
+
+  React.useEffect(() => {
+    addSubscription(serviceContext.target.target().subscribe((_) => updateKey()));
+  }, [addSubscription, serviceContext, updateKey]);
+
+  React.useEffect(() => {
+    addSubscription(controllerContext.controller.hasActiveRecording().subscribe(setHasRecording));
+  }, [addSubscription, controllerContext, setHasRecording]);
+
+  React.useEffect(() => {
+    addSubscription(serviceContext.api.grafanaDashboardUrl().subscribe(setDashboardUrl));
+  }, [addSubscription, serviceContext, setDashboardUrl]);
+
+  React.useEffect(() => {
+    if (!dashboardUrl) {
+      return;
+    }
+    const u = new URL('/d-solo/main', dashboardUrl);
+    u.searchParams.append('theme', props.theme);
+    u.searchParams.append('panelId', String(kindToId(props.chartKind)));
+    u.searchParams.append('to', 'now');
+    u.searchParams.append('from', `now-${props.duration}s`);
+    u.searchParams.append('refresh', `${props.period}s`);
+    setChartSrc(u.toString());
+  }, [dashboardUrl, props.theme, props.chartKind, props.duration, props.period, setChartSrc]);
+
+  React.useEffect(() => {
+    addSubscription(controllerContext.controller.attach().subscribe(setIdx));
+  }, [addSubscription, controllerContext]);
+
+  const refresh = React.useCallback(() => {
+    controllerContext.controller.requestRefresh();
+  }, [controllerContext]);
+
+  React.useEffect(() => {
+    refresh();
+    addSubscription(interval(props.period * 1000).subscribe(() => refresh()));
+  }, [addSubscription, props.period, refresh]);
+
+  const popout = React.useCallback(() => {
+    if (chartSrc && dashboardUrl) {
+      window.open(chartSrc, '_blank');
+    }
+  }, [chartSrc, dashboardUrl]);
+
+  const cardStyle = React.useMemo(() => {
+    return {
+      height: getHeight(props.chartKind),
+    };
+  }, [props.chartKind]);
+
+  const expandButton = React.useMemo(() => {
+    return (
+      <>
```

Comment:
React fragment is missing a key (React is throwing a warning). Here, we can just remove it. Same for `popoutButton`.